app/controllers/messages_controller.rb

```diff
@@ -0,0 +1,86 @@
+class MessagesController < ApplicationController
+    # before_action :correct_message, only: [:destroy]
+    before_action :set_message, only: %i[ show edit update destroy ]
+    def new
+        @message = Message.new
+    end
+
+    def show
+        @message = Message.find(params[:id])
+        
+        # respond_to do |format|
+        #     format.html 
+        #     format.xml {render :xml => @message}
+        #     # format.json  {render :json => @message}
+        # end
+    end
+    
+    def create
+        @message = Message.new(message_params)
+        @message.thumbnail.attach(params[:message][:thumbnail])
+        @message.video.attach(params[:message][:video])
+        if @message.save
+            reset_session
+            flash[:success] = "Successfully added new message"
+            redirect_to root_path
+        else
+            render 'new'
+        end
+    end
+
+    def index
+        params.each do |key,value|
+            Rails.logger.warn "Param #{key}: #{value}"
+            if key == "tag"
+                @messages = Message.joins(:tags).where('tags.name' => value).group(:id)
```

Comment:
Good solution here!  Extra credit question - can you tell me the (approximate, pseudocode) SQL query that this generates?  Just reply to my comment.

If not thats OK, just trying to get a gague on where you're at with SQL knowledge

Reply:
Good question! I believe that `SELECT "messages" * FROM "messages" INNER JOIN "taggings" WHERE "tag.name" = (insert tag name)` is close to the SQL query. I think it also has to find the message.ID and tag.ID in the "taggings" table, and then does another INNER JOIN with the "tags" table. I don't remember exactly how that works. I'd probably have to do a refresher course on SQL - it's been a while.

Reply:
Great, and where does the `group(:id)` come in?

Reply:
Does it group the messages by message id?